qadence/qinfo_tools/qfi.py

```diff
@@ -0,0 +1,261 @@
+        diff_mode (DiffMode, optional): _description_. Defaults to DiffMode.ad.
+
+    Returns:
+        torch.Tensor: QFI matrix
+    """
+
+    if fm_dict == {}:
+        blocks = primitive_blocks(circuit.block)
+        for block in blocks:
+            params = parameters(block)
+            for p in params:
+                if not p.trainable:
+                    fm_dict[p.name] = torch.tensor([p.value])
+
+    # Get Overlap() model
+    ovrlp_model = _overlap_with_inputs(
+        circuit,
+        var_values,  # type: ignore [arg-type]
+        backend=backend,
+        overlap_method=overlap_method,
+        diff_mode=diff_mode,
+    )
+    ovrlp = ovrlp_model(bra_param_values=fm_dict, ket_param_values=fm_dict)
+
+    # Retrieve variational parameters of the overlap model
+    # Importantly, the vparams of the overlap model are the vparams of the bra tensor,
+    # Which means if we differentiate wrt vparams we are differentiating only wrt the
+    # parameters in the bra and not in the ket
+    vparams = {k: v for k, v in ovrlp_model._params.items() if v.requires_grad}
+
+    # Jacobian of the overlap
+    ovrlp_grad = grad(
+        ovrlp,
+        list(vparams.values()),
+        torch.ones_like(ovrlp),
+        create_graph=True,
+        allow_unused=True,
+    )
+
+    # Hessian of the overlap = QFI matrix
+    n_params = ovrlp_model.num_vparams
+    fid_hess = torch.empty((n_params, n_params))
+    for i in range(n_params):
+        ovrlp_grad2 = grad(
+            ovrlp_grad[i],
+            list(vparams.values()),
+            torch.ones_like(ovrlp_grad[i]),
+            create_graph=True,
+            allow_unused=True,
+        )
+        for j in range(n_params):
+            fid_hess[i, j] = ovrlp_grad2[j]
+
+    return -2 * fid_hess
+
+
+def get_quantum_fisher_spsa(
+    circuit: QuantumCircuit,
+    k: int,
+    var_values: tuple | list | Tensor | None = None,
+    fm_dict: dict[str, Tensor] = {},
+    previous_qfi_estimator: Tensor = None,
+    epsilon: float = 10e-4,
+    beta: float = 10e-3,
+    backend: BackendName = BackendName.PYQTORCH,
+    overlap_method: OverlapMethod = OverlapMethod.EXACT,
+    diff_mode: DiffMode = DiffMode.AD,
+) -> Tensor:
+    """Function to calculate the Quantum Fisher Information (QFI) matrix.
+
+    of the quantum state generated.
+
+    by the quantum circuit. The QFI matrix is calculated as the Hessian of
+    the fidelity of the quantum state
+
+    Args:
+        circuit (QuantumCircuit): _description_
+        backend (BackendName, optional): _description_. Defaults to BackendName.pyq.
+        overlap_method (OverlapMethod, optional): _description_. Defaults to OverlapMethod.EXACT.
+        diff_mode (DiffMode, optional): _description_. Defaults to DiffMode.ad.
+
+    Returns:
+        torch.Tensor: QFI matrix
+    """
+    num_vparams = len(var_values) if var_values is not None else 0
```

Comment:
I added the if/else condition depending on type. When would you have None parameters and should it make sense to add a better checking?

Reply:
`var_values` are the values of the variational parameters where you want to compute your QFI matrix. When they are set to `None`, the QFI is calculated in with the default values for the vparams in the model (for example, if you call it after training your model and you want to compute the QFI with the values of the optimal vparams, you can just set `var_values` to None and it will automatically compute it there). However, if you want to calculate the QFI in another point of your landscape, you can pass those values in the input.

Reply:
OK, thanks for the explanation, but the original len(None) in these cases wouldn't work or am I missing something?